ci/download_dependencies.sh

```diff
@@ -0,0 +1,131 @@
+
+source "${WORKDIR}/ci/archs.sh"
+
+temp_dir="${WORKDIR}/bin/deps/releases"
+lib_root="${WORKDIR}/bin/deps/lib"
+
+libtelio_version="v5.0.0-rc2"
+libtelio_url="${LIBTELIO_RELEASES_URL}/${libtelio_version}/downloads/linux.zip"
+libtelio_zipfile="${temp_dir}/libtelio-${libtelio_version}.zip"
+libtelio_dst="${temp_dir}/libtelio-${libtelio_version}"
+
+libdrop_version="v7.0.0"
+libdrop_url="${LIBDROP_RELEASES_URL}/${libdrop_version}/downloads/linux.zip"
+libdrop_zipfile="${temp_dir}/libdrop-${libdrop_version}.zip"
+libdrop_dst="${temp_dir}/libdrop-${libdrop_version}"
+
+libmoose_nordvpnapp_version="v11.1.0-nordVpnApp"
+libmoose_nordvpnapp_url="${LIBMOOSE_RELEASES_URL}/${libmoose_nordvpnapp_version}/downloads/linux.zip"
+libmoose_nordvpnapp_zipfile="${temp_dir}/libmoose-nordvpnapp-${libmoose_nordvpnapp_version}.zip"
+libmoose_nordvpnapp_dst="${temp_dir}/libmoose-nordvpnapp-${libmoose_nordvpnapp_version}"
+
+libmoose_worker_version="v10.0.4-worker"
+libmoose_worker_url="${LIBMOOSE_RELEASES_URL}/${libmoose_worker_version}/downloads/linux.zip"
+libmoose_worker_zipfile="${temp_dir}/libmoose-worker-${libmoose_worker_version}.zip"
+libmoose_worker_dst="${temp_dir}/libmoose-worker-${libmoose_worker_version}"
+
+mkdir -p "${temp_dir}"
+
+function fetch_gitlab_release() {
+  if [[ $# -ne 2 ]]; then
+    echo "Two parameters are required for fetch_gitlab_release function."
+    exit 1
+  fi
+  release_url="${1}"
+  out_file="${2}"
+
+  echo "Downloading release from ${release_url} to ${out_file}"
+  # disable tracing - don't show the token
+  set +x
+  curl \
+    --retry 3 \
+    --retry-delay 2 \
+    --header "PRIVATE-TOKEN: ${PERSONAL_ACCESS_TOKEN}" \
+    -o "${out_file}" \
+    -L "${release_url}"
+  # re-enable tracing
+  set -x
+}
+
+# This function makes a copy of .so file. It requires two parameters:
+# - so_prefix - path to the .so file in unzipped release directory up to architecture name
+# - so_file - name of the .so file to copy
+#
+# Those two parameters are joined with architecture and all together create
+# full path. Example:
+#
+# `copy_to_libs "${libtelio_dst}/linux/release" "libtelio.so"`
+#
+# makes a copy of ${libtelio_dst}/linux/release/${architecture}/libtelio.so to the
+# `libroot/${architecture}/latest directory`.
+function copy_to_libs() {
+  if [[ $# -ne 2 ]]; then
+    echo "Two parameters are required for copy_to_libs function."
+    exit 1
+  fi
+  so_prefix="${1}"
+  so_file="${2}"
+  for arch in "${!ARCHS_SO_REVERSE[@]}"; do
+    so_path="${so_prefix}/${arch}/${so_file}"
+    # libraries have different names for the same architectures and
+    # ARCHS_SO_REVERSE contains architecture names used in all three libraries,
+    # so I check if the .so file exists.
+    if [[ -e ${so_path} ]]; then
+      cp "${so_path}" "${lib_root}/${ARCHS_SO_REVERSE[${arch}]}/latest"
+    fi
+  done
+}
+
+# Release zips are pretty big. Skip downloading if it was already done.
+if [[ -e "${lib_root}/checkout-completed-flag" ]]; then
```

Comment:
Is this optimization for local/mage builds? I assume that in pipelines this script will be ran once at the beginning stages. Would be nice to have an option to run a clean build(always download fresh artifacts). We have the same approach for building the deb package, there are mage targets with `Fast` in the name which skip building if package is already present and targets without `Fast` always rebuild. But maybe just adding(in a separate task) something like a `Clean` target that would remove all the artifacts would be better.

Reply:
Yes, that's optimization for local builds. In the pipeline the build will always be fresh one and it will always download those dependencies with each pipeline run.

Regarding the `Fast` and not-`Fast` builds, I need to think about this, because currently, those release zips and .so files stuff all happen inside `./bin` which gets cleared with `mage clean`. And by default all builds are `Fast` (so they are not re-downloading the release files) after first build. Then the not-`Fast` mode is basically `mage clean && mage build:XYZ` but it forces us to skip everything - currently it's just data build which is additionally skipped because dependencies download happens early, so I'm not sure adding separate `Fast` mode is worth it :thinking: , but on the other hand I like consistency :smile:. I'm leaning towards your suggestion, but I'm not sure. I'll ask the team and we'll see :smile:

Reply:
On the other hand, the current solution is consistent with openvpn download :sweat_smile: - it also doesn't have `Fast` mode and skips the download if it already happened and the cleaning happens via `mage clean` because it's also inside `./bin`

Reply:
We discussed this on the meeting and we are going with the current implementation. There will be work, in a separate ticket, to make it consistent